### ISSUE TITLE
Clang accepts invalid `static_cast` from derived type to private base type in case of rvalue reference

### ISSUE THREAD

**Issue by wangbo15** [llvm/llvm-project#121429]
Clang does not report any error for the following code:

```c++
struct A { int a;};
struct X : private A{};

int main() {
    X x;
    int a=static_cast<A&&>(x).a;
}
```

Please note that this can be correctly rejected if the target type is not a rvalue reference.

GCC, EDG, and MSVC reject it:

```c++
<source>: In function 'int main()':
<source>:6:29: error: 'A' is an inaccessible base of 'X'
    6 |     int a=static_cast<A&&>(x).a;
      |  
```

Please see https://godbolt.org/z/aW1s6a7ec

**Comment by safocl** [comment]
`A program is ill-formed if it uses static_cast to perform the inverse of an ill-formed standard conversion sequence.`
([here](https://timsong-cpp.github.io/cppwp/n4950/expr.static.cast#7))
